### PR TITLE
[ARCHETYPE-584] fix indentation in root pom.xml

### PR DESCRIPTION
**Pull request by newur**
Problem: linebreaks + indentions (whitespaces) are read as actual document nodes. This causes problems in combination with
```
tr.setOutputProperty( OutputKeys.INDENT, "yes" );
tr.setOutputProperty( "{http://xml.apache.org/xslt}indent-amount", "2" );
```

Removed the `.normalizeWhitespace()` in tests, because this shadowed the actual bug. Also split  one test into multiple tests. Test content remains the same.

Had to add the **maven-4.0.0.xsd**, not sure if this is desired.

### CODE REVIEW

**Comment by softdays** [comment]
Any chance to get this PR merged?

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java [hunk 1]
```diff
@@ -239,93 +250,118 @@ private ClassLoader getContextClassloader( Artifact archetypeArtifact, ArtifactR
...
+    public void testAddModuleToParentWithModelVersion()
+            throws Exception
+    {
+        String pom = PROJECT_ELEMENT
+                + MODEL_VERSION
+                + PACKAGING
+                + PROJECT_ELEMENT_END;
 
-        out = new StringWriter();
+        StringWriter out = new StringWriter();
```
Comment (elharo) [inline archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:280]: Looks like this could be a field that doesn't need to be initialized in each method.

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java [hunk 2]
```diff
@@ -61,6 +61,17 @@
 public class ArchetypeTest
     extends PlexusTestCase
 {
+    private static final String XML_VERSION_1_0 =
```
Comment (elharo) [inline archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:64]: splitting out these constants makes the tests harder for me to read. I prefer having all the expected content directly in each method, even at the expense of some duplication.

Comment (newur) [reply archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java]: Funny how different people read code. I could barley make any sense out of the initial version. Is revert in the new commit.

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java (hunk 1 above)
Comment (newur) [reply archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java]: Moved to setup method.

**Comment by newur** [comment]
Any progress on this PR?

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java [hunk 3]
```diff
@@ -61,19 +61,11 @@
...
-            "xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" " +
-            "xsi:schemaLocation=\"http://maven.apache.org/POM/4.0.0 " +
-            "http://maven.apache.org/xsd/maven-4.0.0.xsd\">\n";
-    private static final String PROJECT_ELEMENT_END = "</project>";
-    private static final String MODEL_VERSION = "  <modelVersion>4.0.0</modelVersion>\n";
-    private static final String PACKAGING = "  <packaging>pom</packaging>\n";
 
     private OldArchetype archetype;
 
+    private StringWriter out;
```
Comment (elharo) [inline archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:67]: initialize this here

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java [hunk 4]
```diff
@@ -253,103 +245,139 @@ private ClassLoader getContextClassloader( Artifact archetypeArtifact, ArtifactR
...
+        String pom = ("<project " +
+                "xmlns=\"http://maven.apache.org/POM/4.0.0\" " +
+                "xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" " +
+                "xsi:schemaLocation=\"http://maven.apache.org/POM/4.0.0 " +
+                "http://maven.apache.org/xsd/maven-4.0.0.xsd\">\n")
+                + "  <packaging>pom</packaging>\n"
+                + "</project>";
 
-        StringWriter out = new StringWriter();
+        out = new StringWriter();
```
Comment (elharo) [inline archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:256]: you don't need to initialize this in the test

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java [hunk 5]
```diff
@@ -253,103 +245,139 @@ private ClassLoader getContextClassloader( Artifact archetypeArtifact, ArtifactR
...
+                "xmlns=\"http://maven.apache.org/POM/4.0.0\" " +
+                "xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" " +
+                "xsi:schemaLocation=\"http://maven.apache.org/POM/4.0.0 " +
+                "http://maven.apache.org/xsd/maven-4.0.0.xsd\">\n")
+                + "  <modelVersion>4.0.0</modelVersion>\n"
+                + "  <packaging>pom</packaging>\n"
+                + "</project>";
 
-        StringWriter out = new StringWriter();
+        out = new StringWriter();
```
Comment (elharo) [inline archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:284]: you don't need to initialize this in the test

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java [hunk 6]
```diff
@@ -253,103 +245,139 @@ private ClassLoader getContextClassloader( Artifact archetypeArtifact, ArtifactR
...
+                "http://maven.apache.org/xsd/maven-4.0.0.xsd\">\n")
+                + "  <modelVersion>4.0.0</modelVersion>\n"
+                + "  <packaging>pom</packaging>\n"
                 + "  <modules>\n"
                 + "  </modules>\n"
-                + PROJECT_ELEMENT_END;
+                + "</project>";
 
-        StringWriter out = new StringWriter();
+        out = new StringWriter();
```
Comment (elharo) [inline archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:315]: you don't need to initialize this in the test

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java [hunk 7]
```diff
@@ -253,103 +245,139 @@ private ClassLoader getContextClassloader( Artifact archetypeArtifact, ArtifactR
...
+                + "  <modelVersion>4.0.0</modelVersion>\n"
+                + "  <packaging>pom</packaging>\n"
                 + "  <modules>\n"
                 + "    <module>myArtifactId3</module>\n"
                 + "  </modules>\n"
-                + PROJECT_ELEMENT_END;
+                + "</project>";
 
-        StringWriter out = new StringWriter();
+        out = new StringWriter();
```
Comment (elharo) [inline archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:347]: you don't need to initialize this in the test

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java [hunk 8]
```diff
@@ -253,103 +245,139 @@ private ClassLoader getContextClassloader( Artifact archetypeArtifact, ArtifactR
...
                 + "  <modelVersion>4.0.0</modelVersion>\n"
                 + "  <packaging>pom</packaging>\n"
                 + "  <modules>\n"
                 + "    <module>myArtifactId3</module>\n"
                 + "  </modules>\n"
-                + PROJECT_ELEMENT_END;
+                + "</project>";
 
-        StringWriter out = new StringWriter();
+        out = new StringWriter();
```
Comment (elharo) [inline archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:380]: you don't need to initialize this in the test

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java [hunk 9]
```diff
@@ -381,15 +413,19 @@ public void testAddModuleToParentWithProfiles()
                 + "      </modules>\n"
                 + "    </profile>\n"
                 + "  </profiles>\n" +
-                PROJECT_ELEMENT_END;
+                "</project>";
 
-        StringWriter out = new StringWriter();
+        out = new StringWriter();
```
Comment (elharo) [inline archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:418]: you don't need to initialize this in the test

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java [hunk 10]
```diff
@@ -453,5 +489,6 @@ protected void setUp()
     {
         super.setUp();
         archetype = (OldArchetype) lookup( OldArchetype.ROLE );
+        out = new StringWriter();
```
Comment (elharo) [inline archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:492]: setUp method usually goes near the top

Comment (newur) [reply archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:492]: I kept the method where it was before my patch. I think that keeps reviewing changes easier.
While I personally also prefer it at the top, there is another class which has it after the test, so I assume this is somehow an intended style.

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java (hunk 9 above)
Comment (newur) [reply archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java]: You are absolutly right. Fixed all occurances.

**Comment by newur** [comment]
Updated the branch

**Comment by hboutemy** [comment]
@elharo is this ok now to merge, or do you think something additional is required?

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java (hunk 3 above)
Comment (elharo) [reply archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:67]: ping

Comment (newur) [reply archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:67]: It is initialized in the setup method. If I init it here it would be done twice. If I remove it from the setup the tests would share state.

Or do I miss something here?

Btw if any maintainer _really_ wants it to be changes, please fell free to do so. Whatever works best, I am fine with it.

Comment (elharo) [reply archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:67]: Yes, you're missing something. Tests do not share instance variables. Every test is run with a new object.

Comment (newur) [reply archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java]: Done.

**Comment by hboutemy** [comment]
can't merge, there are 4 ITs failing: https://builds.apache.org/blue/organizations/jenkins/maven-box%2Fmaven-archetype/detail/ARCHETYPE-584/14/pipeline :(

**Comment by jorgheymans** [comment]
> IT failures look real. I'm guessing that the golden data needs to be updated to match the output after this PR:

How does one update the golden data then ? This PR has gone stale but seems really only this last hurdle is missing.

**Comment by newur** [comment]
Wow, this branch/bug is still around. I totally forgot about it.

After playing around with the integration tests a bit, one problem was to load the `.xsd` from the packaged `.jar` (which works differently from reading from a single file). This seems to fix the ITs.

Let's see what comes up next or if this long journey reaches its happy end. :)

File: archetype-common/src/main/java/org/apache/maven/archetype/common/util/PomUtils.java [hunk 11]
```diff
@@ -79,6 +83,13 @@ public static boolean addNewModule( String artifactId, Reader fileReader, Writer
         dbf.setXIncludeAware( false );
         dbf.setExpandEntityReferences( false );
 
+        InputStream inputStream = PomUtils.class.getClassLoader().getResourceAsStream( "maven-4.0.0.xsd" );
```
Comment (elharo) [inline archetype-common/src/main/java/org/apache/maven/archetype/common/util/PomUtils.java:86]: What is the purpose of loading this schema?

File: archetype-common/src/main/resources/maven-4.0.0.xsd [hunk 12]
```diff
@@ -0,0 +1,2246 @@
+<?xml version="1.0"?>
```
Comment (elharo) [inline archetype-common/src/main/resources/maven-4.0.0.xsd:1]: strange that this has to be copied into this project. I'd expect it to be loaded from maven core, if at all.

File: archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java [hunk 13]
```diff
@@ -239,92 +242,152 @@ private ClassLoader getContextClassloader( Artifact archetypeArtifact, ArtifactR
...
 
-    public void testAddModuleToParentPOM()
-        throws Exception
+    public void testAddModuleToParentBasic()
+            throws Exception
     {
-        String pom = "<project>\n"
-            + "  <packaging>pom</packaging>\n"
-            + "</project>";
+        String pom = ("<project " +
```
Comment (elharo) [inline archetype-common/src/test/java/org/apache/maven/archetype/old/ArchetypeTest.java:248]: This is a good fix, but does not appear related to the description of this PR

File: archetype-common/src/main/java/org/apache/maven/archetype/common/util/PomUtils.java (hunk 11 above)
Comment (newur) [reply archetype-common/src/main/java/org/apache/maven/archetype/common/util/PomUtils.java:86]: Please see my minimal, reproducible example here https://issues.apache.org/jira/browse/ARCHETYPE-584?focusedCommentId=17008264&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17008264

I investigated this 1.5 years ago and can not remember the details why it only worked this way. However, I am pretty sure that I did not dive into the implementation details of `DocumentBuilderFactory`, which is fed with the `schema`, so not sure if I would have been able to explain it very deep back then.

File: archetype-common/src/main/resources/maven-4.0.0.xsd (hunk 12 above)
Comment (newur) [reply archetype-common/src/main/resources/maven-4.0.0.xsd:1]: Feel free to point me to location of the `xsd` in maven core or any other way to load it and I happily update the PR.

File: archetype-common/src/main/java/org/apache/maven/archetype/common/util/PomUtils.java (hunk 11 above)
Comment (elharo) [reply archetype-common/src/main/java/org/apache/maven/archetype/common/util/PomUtils.java:86]: My guess is that you don't need the schema to fix this bug. In fact, you probably only need `setIgnoringElementContentWhitespace` though `setNamespaceAware( true )` is always a good idea.

Comment (newur) [reply archetype-common/src/main/java/org/apache/maven/archetype/common/util/PomUtils.java:86]: Please check the [example](https://issues.apache.org/jira/browse/ARCHETYPE-584?focusedCommentId=17008264&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17008264) I provided. You will see that without the schema the output is not as expected.

**Comment by newur** [comment]
I committed a new attempt to get this merged. Since I do not understand which solution is acceptable, I made an unclean commit (code in comments, no proper exception handling).
Would be nice to have some feedback, if this solution has any chance to get accepted before I clean it up.

**Comment by kwin** [comment]
I think this is superseded by #128

**Comment by newur** [comment]
I think so too and would love to try it out. Unfortunately maven-archetype-plugin version 3.3.0 is not released, yet.

Any estimated time when the release will happen? Or any way to test this, besides building the plugin locally?

**Comment by Giovds** [comment]
Hi @newur, I just randomly came across this PR. I just wanted to let you know the maven-archetype-plugin `3.3.0` has been released since 22 Sep 2024. Perhaps you can retest with that version to verify if the issue is still around

**Comment by newur** [comment]
Hello @Giovds, I tested it with `3.3.0` and indeed the line break issue is gone. Works as expected.

Thanks for the notification.

**Comment by jira-importer** [comment]
Resolve #818

**Comment by jira-importer** [comment]
Resolve #818
